### PR TITLE
Implement tls-verify switch in libpod API

### PR DESCRIPTION
**Pull request by rhatdan**
Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

#### Does this PR introduce a user-facing change?

```release-note
--tls-verify switch now works with podman -remote
```

### CODE REVIEW

**Comment by TomSweeneyRedHat** [comment]
LGTM
but ugly red tests

**Comment by SoMuchForSubtlety** [comment]
As far as I can tell, this PR does no change the default behaviour of not allowing pulling from http, correct?

**Comment by vrothberg** [comment]
> As far as I can tell, this PR does no change the default behaviour of not allowing pulling from http, correct?

The default behavior must not change.  Docker doesn't disable TLS verification by default either as it would be a huge security risk.

File: pkg/api/handlers/compat/images.go [hunk 1]
```diff
@@ -259,10 +259,12 @@ func CreateImageFromImage(w http.ResponseWriter, r *http.Request) {
 
 	query := struct {
 		FromImage string `schema:"fromImage"`
-		Tag       string `schema:"tag"`
 		Platform  string `schema:"platform"`
+		TLSVerify bool   `schema:"tlsVerify"`
```
Comment (vrothberg) [inline pkg/api/handlers/compat/images.go:263]: I don't see the Docker API having that parameter.

**Comment by SoMuchForSubtlety** [comment]
Looks like docker has a special case for localhost hardcoded
```go
// loadInsecureRegistries loads insecure registries to config
func (config *serviceConfig) loadInsecureRegistries(registries []string) error {
	// Localhost is by default considered as an insecure registry. This is a
	// stop-gap for people who are running a private registry on localhost.
	registries = append(registries, "127.0.0.0/8")
```

https://github.com/moby/moby/blob/6eab4f55faf77bb0f253553d7d6fadb2ef6f4669/registry/config.go#L182-L186

**Comment by vrothberg** [comment]
Great catch, @SoMuchForSubtlety!

@mtrmac WDYT? Should c/image follow that example?

**Comment by mtrmac** [comment]
- I’m unsure about excluding `localhost`. What about network namespaces, multi-user systems, and the like? Accessing `localhost` is not _obviously_ not crossing a security boundary, although it’s certainly likely that most uses will not [just like most TLS connections are not under attack…]
- We _definitely must not_ use the `InsecureRegistryCIDRs` approach of a separate DNS lookup; that’s a TOCTOU that allows bypassing TLS entirely. If anything, a plain text match on `localhost`.
    - So, we would not be _truly compatible_ either way.
    - Last time I have checked, which was quite a few years ago, there really was no way to avoid the separate DNS lookup / TOCTOU in the Go network / HTTP stack. Maybe that has changed.
    - (Users can already disable TLS for the `localhost` name in `registries.conf`. Admittedly that’s not an answer to the question of whether we should do that by default.)

**Comment by mtrmac** [comment]
It seems to me that if we should recommend that _without qualification_, we should also ship that configuration by default … which brings the question full circle.

Alternatively, recommend it for single-user workstations; that would probably be more correct but I don’t know where we would document it so that people read it.

**Comment by SoMuchForSubtlety** [comment]
> Should we recommend configuring `registries.conf` instead? I think that' acceptable to turn testcontainers-CI green.

It's currently not possible to allow localhost without specifying a port.

The following config yields the error `invalid condition: location is unset and prefix is not in the format: *.example.com"`
```toml
[[registry]]
prefix = "localhost"
insecure = true
```
This config works, but only for the specified port, and the tests use randomly assigned ports.
```toml
[[registry]]
location = "localhost:123123"
insecure = true
```

**Comment by mtrmac** [comment]
Huh?
```
[[registry]]
location = "localhost"
insecure = true
```

is expected to work, and seems to work for me manually modifying the existing unit tests. If it doesn’t work, please file a c/image bug with a reproducer.

**Comment by SoMuchForSubtlety** [comment]
> is expected to work, and seems to work for me manually modifying the existing unit tests. If it doesn’t work, please file a c/image bug with a reproducer.

seems to be a known issue? 
https://github.com/containers/image/blob/0d85878d7a774dd753a4e211455282af94d78556/pkg/sysregistriesv2/system_registries_v2.go#L392-L400
```go
// FIXME: allow config authors to always use Prefix.
// https://github.com/containers/image/pull/1191#discussion_r610622495
if !strings.HasPrefix(reg.Prefix, "*.") && reg.Location == "" {
	return &InvalidRegistries{s: "invalid condition: location is unset and prefix is not in the format: *.example.com"}
}
```

**Comment by vrothberg** [comment]
@mtrmac set the `location` while the issue happens when only "prefix` is set.  What doesn't work is to set _any_ port on localhost as insecure.

**Comment by mtrmac** [comment]
> What doesn't work is to set any port on localhost as insecure.

Good point, that’s true.

**Comment by mtrmac** [comment]
… and that weakens my argument that users can already set this in `registries.conf`. It’s kind of true but it’s noticeably more of a hassle for temporary per-user registries.

**Comment by mtrmac** [comment]
Very generally speaking there the concern there is maintaining a comprehensible “closest match” ordering between the registry entries. In this instance, I think a new `localhost:*` syntax could work — _but_ note that the implementation (sorting `refMatchingPrefix` matches by `len(r.Prefix)` would need changing to explicitly order `localhost:*` vs `localhost/a` correctly.

I don’t know that this resolves the question about having a built-in default — but making that config _possible_ seems with a `localhost:*` entry might be valuable either way.

**Comment by vrothberg** [comment]
@mtrmac since it's a special case one way or another: Would do you think of  a boolean option "unsafe_localhost_insecure=true"?

**Comment by mtrmac** [comment]
(`localhost-no-tls`? maybe we don’t need to warn so strongly in this case, and “insecure” is unspecific about this being TLS.)

Where would that option live? As a special case in `registries.conf`? Right now I’d mildly prefer to have the more generic `$host:*` feature in that case…

But really I’m uninformed and undecided at this point. Should this be a built-in default? I’m almost sure we have had that discussion before, both about `localhost` special cases, and about host:anyPort matching; I can’t remember a thing about any upsides/downsides discussed. Given enough time, I will research this eventually…

**Comment by vrothberg** [comment]
> (`localhost-no-tls`? maybe we don’t need to warn so strongly in this case, and “insecure” is unspecific about this being TLS.)

SGTM

> Where would that option live? As a special case in `registries.conf`? Right now I’d mildly prefer to have the more generic `$host:*` feature in that case…

Yes, I had registries.conf in mind.
 
> But really I’m uninformed and undecided at this point. Should this be a built-in default? I’m almost sure we have had that discussion before, both about `localhost` special cases, and about host:anyPort matching; I can’t remember a thing about any upsides/downsides discussed. Given enough time, I will research this eventually…

I had a simple boolean in mind.  False unless set in registries.conf.  The code treating insecure can deal with it if the domain is localhost.

A general host:anyPort would certainly be cool but I am concerned about the time it may consume.  I think our books are quite full at the moment.

**Comment by vrothberg** [comment]
Closing but we should continue the design discussion and then move the issue over to c/image.

**Comment by rhatdan** [comment]
Why is this closed from a Podman point of view.

Currently 
podman pull --tls-verify=false  works,
podman --remote pull --tls-verify does not?

**Comment by vrothberg** [comment]
> Why is this closed from a Podman point of view.

It claims to fix https://github.com/containers/podman/issues/16486 but is the wrong approach. 

> Currently podman pull --tls-verify=false works, podman --remote pull --tls-verify does not?

Shouldn't that change the libpod endpoint? This PR changes the compat one.

**Comment by rhatdan** [comment]
Ok I will reopen to fix the libpod endpoint.

**Comment by vrothberg** [comment]
@rhatdan shall we close?

**Comment by rhatdan** [comment]
No I hope to get back to this one.

File: pkg/api/handlers/compat/images.go (hunk 1 above)
Comment (vrothberg) [reply pkg/api/handlers/compat/images.go:264]: @rhatdan the Docker API doesn't have such a parameter.  I don't think this PR is needed.

**Review by vrothberg — COMMENTED** [review]
@rhatdan, the `compat` endpoint does not support a `tlsVerify` parameter.  `podman-remote` uses the `libpod` specific `/images/pull` endpoint.

I don't think the code change has an effect.

**Pull request closed without merging**